pool/flexpool.go

```diff
@@ -0,0 +1,77 @@
+package pool
+
+import (
+	"sync"
+)
+
+type FlexPool struct {
+	mu      sync.Mutex
+	taskQueue map[string]chan func()
+	useCount map[string]int
+	terminating bool
+}
+
+func NewFlexPool() *FlexPool {
+	// taskQueue is unbuffered since items are always removed immediately.
+	pool := &FlexPool{
+		taskQueue:   make(map[string]chan func()),
+		useCount: make(map[string]int),
+	}
+	return pool
+}
+
+func (p *FlexPool) Terminate() {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	if p.terminating {
+		return
+	}
+	p.terminating = true
+	for _, q := range p.taskQueue {
+		close(q)
+	}
+}
+
+func (p *FlexPool) IsTerminating() bool {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	return p.terminating
+}
+
+func (p *FlexPool) Submit(uid string, task func()) {
+	if p.IsTerminating() || task == nil {
+		return
+	}
+	p.mu.Lock()
```

Comment:
Calling `IsTerminating` and then acquiring the lock is a race -- the pool can be terminated between those calls. We should probably acquire the lock, check `p.terminating`, and then proceed accordingly.